src/InputParser.js

```diff
@@ -0,0 +1,81 @@
+import { CUSTOM_DELIMITER, ERROR, GROUPING_REGEX, NEED_ESCAPE_CHARS } from './constants.js';
+class InputParser {
+  #delimiterSet;
+
+  constructor() {
+    this.#delimiterSet = new Set([',', ':']);
+  }
+
+  parse(input) {
+    const { customDelimiters, expression } = InputParser.#getParts(input);
+    this.#addDelimiters(customDelimiters);
+    const numbers = this.#expressionToNumber(expression);
+
+    return numbers;
+  }
+
+  static #getParts(input) {
+    const { customDelimiters, expression } = input.match(GROUPING_REGEX).groups;
+    if (
+      !customDelimiters
+      && (input.startsWith(CUSTOM_DELIMITER.START)
+      || input.includes(CUSTOM_DELIMITER.END))
+    ) {
+      throw new Error(`${ERROR.HEADER} 커스텀 구분자의 형식이 바르지 않습니다.`);
+    }
+
+    return { customDelimiters, expression };
+  }
+
+  #addDelimiters(customDelimiter) {
+    if (!customDelimiter) {
+      return;
+    }
+    if (customDelimiter.length === 0) {
+      throw new Error(`${ERROR.HEADER} ${ERROR.MISSING_CUSTOM_DELIMITER}`);
+    }
+    Array.from(customDelimiter).forEach(this.#addDelimiter.bind(this));
+  }
+
+  #addDelimiter(delimiter) {
+    if (NEED_ESCAPE_CHARS.some(char => char === delimiter)) {
+      return this.#delimiterSet.add(`\\${delimiter}`);
+    }
+
+    return this.#delimiterSet.add(delimiter);
+  }
+
+  #expressionToNumber(expresssion) {
+    if (!expresssion) {
+      return [0];
+    }
+    const splitedExpression = expresssion.split(this.#getSplitRegex());
+    const numbers = splitedExpression.reduce((result, curr) => {
+      if (curr === '') {
+        return result;
+      }
+      const num = InputParser.#parseNumber(Number(curr));
+
+      return [...result, num];
+    }, []);
+
+    return numbers;
+  }
```

Comment:
`expressionToNumber` 함수의 동작이 조금 복잡한 거 같아서 이해가 쉽지 않은 거 같아요 조금 더 작은 함수 단위로 분리할 수는 없을까요?

Reply:
저도 코드를 다시보면서 이부분에서 reduce의 콜백때문에 코드르 좀더 나눴어야 싶더라구요. reduce의 return이 많다보니까 #expressionToNumber가 조금 복잡해 보이나 싶었습니다.
```suggestion
  #expressionToNumber(expresssion) {
    if (!expresssion) {
      return [0];
    }
    const splitedExpression = expresssion.split(this.#getSplitRegex());
    const numbers = splitedExpression.reduce(this.#reduceToNumbers);
    
    return numbers;
  }
```
이런식으로 분리 한다면 좀 나아질까요?:)